ARCHITECTURE.md

```diff
@@ -0,0 +1,86 @@
+# How Riff uses Nix
+
+Riff is written in [Rust] but powered by the [Nix] package manager. This doc is
+mostly conceptual and intended for those who are curious about how Riff uses Nix
+while exposing very little of Nix to users. If you want to know more about how
+to *use* Riff, we recommend consulting the project [README].
+
+## Language support
+
+At the moment, Riff supports [Rust] projects. We intend to provide support for
+additional languages in the future, so stay tuned for updates from us.
+
+## How Riff uses Nix flakes
+
+[Flakes][flake] are an opt-in experimental feature for Nix that enable you to
+encapsulate your Nix dependencies and outputs&mdash;packages, dev environments,
+libraries, and more&mdash;in a standardized and declarative way.
+
+Riff uses an internal [template] to generate a [`flake.nix`][flake] file tailored to
+the specific needs of your Rust project. To build that `flake.nix`, Riff [traverses
+your Rust project's dependency graph](#package-metadata), using the [`cargo
+metadata`][cargo metadata] command, and supplies the necessary external
+dependencies to the shell's `buildInputs` (the list of packages that are
+included in the Nix shell environment).
+
+Once Riff has generated its internal flake, the `riff shell` command essentially
+wraps the [`nix develop`][nix develop] command while `riff run` wraps `nix
+develop --command`. The `flake.nix` file itself, however, isn't written to disk
+and thus doesn't end up in your project directory.
```

Comment:
It could be written to disk, depending on the user's setup... the real distinction is that it's written to a _temporary directory_.